src/pages/ServiceDetails/ServiceMetrics.tsx

```diff
@@ -114,19 +103,17 @@ class ServiceMetrics extends React.Component<ServiceMetricsProps, ServiceMetrics
     API.getServiceMetrics(authentication(), this.props.namespace, this.props.service, this.options)
       .then(response => {
         const metrics: M.Metrics = response.data;
-        this.setState({
-          requestCountIn: metrics.metrics['request_count_in'],
-          requestCountOut: metrics.metrics['request_count_out'],
-          requestSizeIn: metrics.histograms['request_size_in'],
-          requestSizeOut: metrics.histograms['request_size_out'],
-          requestDurationIn: metrics.histograms['request_duration_in'],
-          requestDurationOut: metrics.histograms['request_duration_out'],
-          responseSizeIn: metrics.histograms['response_size_in'],
-          responseSizeOut: metrics.histograms['response_size_out'],
-          tcpReceivedIn: metrics.metrics['tcp_received_in'],
-          tcpReceivedOut: metrics.metrics['tcp_received_out'],
-          tcpSentIn: metrics.metrics['tcp_sent_in'],
-          tcpSentOut: metrics.metrics['tcp_sent_out']
+        this.setState(prevState => {
+          Object.keys(prevState.charts).forEach(k => {
+            const chart = prevState.charts[k];
+            const histo = metrics.histograms[k];
+            if (histo) {
+              chart.metrics = histo;
+            } else {
+              chart.metrics = metrics.metrics[k];
+            }
+          });
+          return prevState;
```

Comment:
IIRC, `prevState` should be used as a constant. Here you are changing some values.
I tried the code and it works. But not sure if React may do weird things under some circumstances.

Reply:
You are right @israel-hdez 
from the [docs](https://reactjs.org/docs/react-component.html#setstate):
> prevState is a reference to the previous state. It should not be directly mutated. Instead, changes should be represented by building a new object based on the input from prevState
> ...
> The output of the updater is shallowly merged with prevState

Reply:
ok I'll fix that